src/tools/miri/src/eval.rs

```diff
@@ -268,10 +268,14 @@ pub fn create_ecx<'tcx>(
     entry_type: EntryFnType,
     config: &MiriConfig,
 ) -> InterpResult<'tcx, InterpCx<'tcx, MiriMachine<'tcx>>> {
-    let param_env = ty::ParamEnv::reveal_all();
-    let layout_cx = LayoutCx::new(tcx, param_env);
-    let mut ecx =
-        InterpCx::new(tcx, rustc_span::DUMMY_SP, param_env, MiriMachine::new(config, layout_cx));
+    let typing_env = ty::TypingEnv::fully_monomorphized();
+    let layout_cx = LayoutCx::new(tcx, typing_env);
+    let mut ecx = InterpCx::new(
+        tcx,
+        rustc_span::DUMMY_SP,
+        typing_env.param_env,
+        MiriMachine::new(config, layout_cx)
+    );
```

Comment:
```suggestion
    );
    assert_eq!(ecx.typing_env(), typing_env);
```
Does this make sense? Currently it seems a bit odd that we only pass part of the typing env to `InterpCx`.

Reply:
The assert should hold, I don't like adding that assert from a style perspective however

a question here is the following:
- const evaluation should always run in the `PostAnalysis/Reveal::All` mode, so we don't *need* to store the `TypingMode` in the interpcx
- we still want to assert that users of const eval pass in the correct `typing_env` and don't accidentally switch to `PostAnalysis` without noticing
- not sure about the final result from my chat with @compiler-errors, but I think we probably want to store a full `TypingEnv` in the `InterpCx` and const eval in general and just assert that the `TypingMode` is `PostAnalysis`. Storing another 16? bytes in the context really shouldn't matter I think

Reply:
>  I don't like adding that assert from a style perspective however

It seems currently there is a non-local invariant encoded by how `typing_env()` is implemented on InterpCx. I'd like to ensure that we realize if this ever breaks.

> we still want to assert that users of const eval pass in the correct typing_env and don't accidentally switch to PostAnalysis without noticing

That sounds reasonable, but that's a different assertion -- that would be somewhere in the const_eval queries. This here is about Miri.

> I think we probably want to store a full TypingEnv in the InterpCx and const eval in general and just assert that the TypingMode is PostAnalysis. Storing another 16? bytes in the context really shouldn't matter I think

I'm fine with that as well.

Reply:
do you mind me changing ` InterpCx` to contain the `typing_env` directly in a followup PR?

Reply:
Yeah feel free to stage this however.